core/system/src/export_table.c

```diff
@@ -83,16 +95,446 @@ static void thread_destroy(void * c)
+    uvisor_rpc_fn_group_t * fn_group_array = index->rpc_fn_group_queue->fn_groups;
+
+    /* Wake up all known waiters for this function. Search for the function in
+     * all known function groups. We have to search through all function groups
+     * (not just those currently waiting for messages) because we want the RTOS
+     * to be able to pick the highest priority waiter to schedule to run. Some
+     * waiters will wake up and find they have nothing to do if a higher
+     * priority waiter already took care of handling the incoming RPC. */
+    uvisor_pool_slot_t slot;
+    slot = fn_group_queue->head;
+    while (slot < fn_group_queue->pool->num) {
+        /* Look for the function in this function group. */
+        uvisor_rpc_fn_group_t * fn_group = &fn_group_array[slot];
+
+        TFN_Ptr const * fn_ptr_array = fn_group->fn_ptr_array;
+        size_t i;
+
+        for (i = 0; i < fn_group->fn_count; i++) {
+            /* If function is found: */
+            if (fn_ptr_array[i] == function) {
+                /* Wake up the waiter. */
+                semaphore_post(&fn_group->semaphore);
+                ++num_posted;
+            }
+        }
+
+        slot = fn_group_queue->pool->management_array[slot].queued.next;
+    }
+
+    return num_posted;
+}
+
+static int callee_box_id(const TRPCGateway * gateway)
+{
+    return (uint32_t *)gateway->box_ptr - __uvisor_config.cfgtbl_ptr_start;
+}
+
+static int put_it_back(uvisor_pool_queue_t * queue, uvisor_pool_slot_t slot)
+{
+    int status;
+    status = uvisor_pool_queue_try_enqueue(queue, slot);
+    if (status) {
+        /* We could dequeue an RPC message, but couldn't put it back. */
+        /* It is bad to take down the entire system. It is also bad
+         * to lose messages due to not being able to put them back in
+         * the queue. However, if we could dequeue the slot
+         * we should have no trouble enqueuing the slot here. */
+        assert(false);
+    }
+
+    /* Note that we don't have to modify the message in the queue, since it'll
+     * still be valid. Nobody else will have run at the same time that could
+     * have messed it up. */
+
+     return status;
+}
+
+/* Return true iff gateway is valid. */
+static int is_valid_rpc_gateway(const TRPCGateway * const gateway)
+{
+    /* Gateway needs to be entirely in flash. */
+    uint32_t gateway_start = (uint32_t) gateway;
+    uint32_t gateway_end = gateway_start + sizeof(*gateway);
+    if (!(vmpu_public_flash_addr(gateway_start) && vmpu_public_flash_addr(gateway_end))) {
+        return 0;
+    }
+
+    /* The callee box ID associated with the gateway must be valid. Gateways to
+     * box 0 are not possible. */
+    int box_id = callee_box_id(gateway);
+    if (box_id <= 0 || box_id >= g_vmpu_box_count) {
+        return 0;
+    }
+
+    /* Gateway needs to have good magic. */
+    if (!(gateway->magic == UVISOR_RPC_GATEWAY_MAGIC_ASYNC || gateway->magic == UVISOR_RPC_GATEWAY_MAGIC_SYNC)) {
```

Comment:
@Patater Make sure to verify fist the magic before you verify members - line 184 with callee_box_id.